[COT] Add the orders cache

Konamiman opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Use [the recently implemented object cache](https://github.com/woocommerce/woocommerce/pull/33634) to cache orders. This includes:

1. A new `OrderCache` class (which is very simple since the bulk of the work is done by the base `ObjectCache` class).
2. A new `Orders cache` experimental feature (uses the recently introduced [features engine](https://github.com/woocommerce/woocommerce/pull/34616)).
3. A new `OrderCacheController` class to control the cache usage, including turning it off temporarily (this is done while synchronization is in progress).
4. Caching all orders that are retrieved via `WC_Order_Factory` (this covers the `wc_get_order(s)` functions, the `WC_Order_Query` class and `query` in the new data store class).
5. Updating the cached value of orders whenever they are saved, but only if the order had been already cached.
6. Dropping orders from the cache whenever they are trashed or deleted (this is done by hooking into `woocommerce_trash_order` and `woocommerce_delete_order`).
7. Flushing the entire cache whenever the custom orders table usage is enabled or disabled.

![image](https://user-images.githubusercontent.com/937723/190623702-314d9c42-39d8-43a5-843e-26e4468fd6e5.png)

A check for the orders cache feature being enabled and not being temporarily disabled is made before any caching operation.

The `OrderCacheController` class includes two methods, `temporarily_disable_orders_cache_usage` and `maybe_restore_orders_cache_usage`. The former stores an in-memory variable with a backup of the cache enable cache setting, the later removes the variable. The feature isn't actually modified, the temporary variable is only used by the `orders_cache_usage_is_enabled` method, which is the one used before doing any order caching operation.

As a bonus, this pull request introduces a couple of tweaks to the base `ObjectCache` class:

* A new `update_if_cached` method.
* `set` will now validate the object to cache _before_ attempting to retrieve its id. This way, if the object is invalid the error returned will be "invalid object" instead of "can't get the id".

Closes #34100.

### How to test the changes in this Pull Request:

In order to ease testing this pull request introduces a temporary transients based cache engine (the `TransientsEngine` class) that not only is persistent unlike the default engine, but also prints a log of all the performed operations in the standard output, this is useful to test the feature using the wp cli tool. **This class must be removed before merging the pull request!**

The transients engine can be activated with the following snippet:

```php
add_filter('wc_object_cache_get_engine', function($engine, $cache) {
	if($cache instanceof \Automattic\WooCommerce\Caches\OrderCache) {
		return wc_get_container()->get(\Automattic\WooCommerce\Caching\TransientsEngine::class);
	}
	return $engine;
}, 10, 2);
```

If you ever want to manually delete the transients created by this engine just run this in you db: `delete from wp_options where option_name like '_transient_%woocommerce_object_cache|order|%'`

The orders cache works with and without the custom orders table feature enabled. Remember that the custom orders table feature is still in development and needs to be explicitly enabled as well (WooCommerce - Settings - Advanced - Features).

With this in place, here's how to tes using wp cli, including the commands to run and the expected response (numbers 100, 101, etc are supposed to be valid ids of orders in the custom table); you should test both with and without the custom orders table feature enabled, and when it's enabled, with and without the new orders table being authoritative:

### Invalid order id

```
$ wp eval 'wc_get_order(99999);'

------- GET: null
Exception caught in get_order. Invalid order.. Args: Array
(
    [0] => 99999
)
```

### Valid order before and after being cached

```
$ wp eval 'wc_get_order(100);'
------- GET: null
------- SET: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|100

$ wp eval 'wc_get_order(100);'
------- GET: cached object retrieved! key: woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|100
```

### Saving an order

Assuming order not previously cached:

```
`$ wp eval 'wc_get_order(100)->save();'
------- GET: null
------- SET: key woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|100
------- GET: cached object retrieved! key: woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|100
------- IS CACHED woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|100? 1
------- SET: key woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|100
```

### get_orders using factory

Note how the code is smart enough to discriminate orders already cached (gets them from cache) and orders not yet cached (gets them and caches them):

```
$ wp eval 'WC_Order_Factory::get_orders([100,101,102]);'
------- GET: null
------- GET: null
------- GET: null
------- SET: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|100
------- SET: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|101
------- SET: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|102

$ wp eval 'WC_Order_Factory::get_orders([103,101,102]);'
------- GET: null
------- GET: cached object retrieved! key: woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|101
------- GET: cached object retrieved! key: woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|102
------- SET: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|103
```

### Trashing and deleting orders

```
$ wp eval 'do_action("woocommerce_trash_order", 100);'
------- DELETE: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|100

$ wp eval 'do_action("woocommerce_delete_order", 100);'
------- DELETE: key woocommerce_object_cache|order|18281ccdabc4f679196b05dd5eb|100
```

### Enabling/disabling the table usage

Assuming the order had been previously cached:

```
$ wp option update woocommerce_custom_orders_table_enabled no
$ wp option update woocommerce_custom_orders_table_enabled yes
$ wp eval 'wc_get_order(100);'
------- GET: null
------- SET: key woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|100
```

### wc_get_orders

Assuming orders already cached:

```
$ wp eval 'wc_get_orders(["return"=>"ids"]);'
(no output since no orders are actually retrieved)

$ wp eval 'wc_get_orders(["limit"=>5]);'
------- GET: cached object retrieved! key: woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|100
------- GET: cached object retrieved! key: woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|101
------- GET: cached object retrieved! key: woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|102
------- GET: cached object retrieved! key: woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|103
------- GET: cached object retrieved! key: woocommerce_object_cache|order|182b07520cbbb3b784b07783d7b|104
```

### Temporarily disabling the cache usage

The cache usage can be temporarily disable programmatically, and it's also disable while syncrhonization is in progress.

To test the programmatic approach:

1. Enable the cache usage in the settings UI.
2. Run the following in a console:

```
$ wp eval '$cc = wc_get_container()->get(Automattic\WooCommerce\Caches\OrderCacheController::class); $cc->temporarily_disable_orders_cache_usage(); wc_get_order(100); $cc->temporarily_disable_orders_cache_usage(); wc_get_order(200);'
```

You should see that the cache is used once (for the order with id 200) and that no cache flush happens.

### Further testing

Create, delete, edit orders via UI and verify that everything is still working as expected.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Reverse the order of "$id" and "$object" in ObjectCache::set

Konamiman committed:
Add the ObjectCache::update_if_cached method

Konamiman committed:
Modify ObjectCache:set to validate object before invoking get_id

Konamiman committed:
Add a temporary TransientsEngine class.

This is temporary! Must be removed before merging to trunk.

Konamiman committed:
Add the OrderCache class.

This class uses a TransientsEngine instance as the caching engine.
This is temporary and must be undone (get_cache_engine_instance method
must be removed) before merging to trunk!

Konamiman committed:
Use the new OrdersCache class

- When an order is retrieved, cache it
- When an order is saved, update it if it was cached already
- When an order is trashed or deleted, remove it from cache
- When the authoritative table for orders changes, flush the cache

Konamiman committed:
Remove the hardcoded usage of TransientEngine in OrderCache

It will make things easier later before merging. The transients engine
can still be used via the wc_object_cache_get_engine hook.

Konamiman committed:
Add changelog file

Konamiman committed:
Merge branch 'trunk' into add/order_cache

vedanshujain reviewed (commented):
Thanks for the PR, looks pretty neat 🚀 . Although, I am wondering if we should link the cache in `WC_Abstract_Order` instead of in order factory, that way we will be able to handle save/update etc properly. Wdyt?

Additionally, I assume that the cot enabled check is only used to reduce the surface area, but otherwise, this would work as it is for posts data store as well. Is that correct?

Konamiman commented:
> Although, I am wondering if we should link the cache in WC_Abstract_Order instead of in order factory, that way we will be able to handle save/update etc properly

Saving is already handled in `WC_Abstract_Order::save`, is there any other place for save/update operations that I'm missing?

If what you mean is that caching should happen in the constructor of `WC_Abstract_Order`, I think that would be a too low-level place for it. There might be cases in which we want to get an order without caching it (unit tests for example), and the only "official" ways to retrieve an order are the order factory and the order query anyway.

In fact I'm even thinking of removing the cache update operation from `WC_Abstract_Order` and handle it in the cache class itself via the `woocommerce_after_order_object_save` event.

>  I assume that the cot enabled check is only used to reduce the surface area, but otherwise, this would work as it is for posts data store as well. Is that correct?

As I understand it the cache engine is, at least for now, part of the COT project, and as such, it should be used only when the COT usage is enabled, at least initially (so you as the store owner either enable COT all the way or none of it). But in principle, yes, this cache is able to handle orders regardless of where they come from.

Konamiman committed:
Merge branch 'trunk' into add/order_cache

Konamiman committed:
Fix failing unit test

The test was failing because the order is cached by reference when
being saved in the test, and then when being deleted by the REST API
code it gets its id set to 0.

Konamiman committed:
Add a setting to enable/disable the orders cache

Also added a mechanism to temporarily disable the orders cache while
syncrhonization is in progress.

Konamiman committed:
Merge branch 'trunk' into add/order_cache

Konamiman committed:
Adjustments in the mechanism to temporarily disable the orders cache usage

vedanshujain commented on plugins/woocommerce/includes/class-wc-order-factory.php:
```diff
@@ -28,13 +31,26 @@ public static function get_order( $order_id = false ) {
 		try {
-			return new $classname( $order_id );
+			$order = new $classname( $order_id );
+			if ( $use_orders_cache ) {
```
We should add an `instance of` check for the order class. Some extensions may return a wp_error object if they fail to initialize the object, and we will end up caching the error object in that case.

Konamiman replied:
You mean before executing `$order_cache->set`? The `OrderCache::validate` method is already checking if the passed object is an `WC_Abstract_Legacy_Order`. That said, when that's not the case an exception is throw, this will cause `get_order` to return false and that's probably not a good idea; so yes, I'll add a check before caching.

vedanshujain commented on plugins/woocommerce/includes/class-wc-order-factory.php:
```diff
@@ -28,13 +31,26 @@ public static function get_order( $order_id = false ) {
+		if ( $use_orders_cache ) {
+			$order_cache = wc_get_container()->get( OrderCache::class );
+			$order       = $order_cache->get( $order_id );
+			if ( ! is_null( $order ) ) {
```
let's also add a `instance of` check, to accommodate for cases where cache engine has some issue and is returning an error object.

vedanshujain replied:
Or perhaps we should add them in the `ObjectCache.php` itself to address similar cases everywhere.

Konamiman replied:
> to accommodate for cases where cache engine has some issue and is returning an error object

This can't happen: the cache engine must either return a cached object or null (that's how the `CacheEngine` interface is defined), issues can be logged but the cache engine classes should never throw exceptions or return error objects.

vedanshujain commented on plugins/woocommerce/src/Caches/OrderCacheController.php:
```diff
@@ -0,0 +1,112 @@
+			return;
+		}
+
+		update_option( self::ORDERS_CACHE_USAGE_ENABLED_BACKUP_OPTION, get_option( self::ORDERS_CACHE_USAGE_ENABLED_OPTION ) );
```
if we are disabling it temporarily, it should probably not be persistently stored in an option. the use case of this method is to disable it in a request, or in a corn job for example. so maybe we can store this in a memory and it should be cleared whenever the request/cron ends.

Konamiman replied:
Makes sense. I'll switch from an option to a property inside the controller class (which is single-instance).

vedanshujain commented on plugins/woocommerce/src/Caching/ObjectCache.php:
```diff
@@ -111,7 +111,7 @@ public function get_default_expiration_value(): int {
 	 */
 	private function get_cache_engine(): CacheEngine {
-		if ( null === $this->cache_engine ) {
+		if ( $this->cache_engine === null ) {
```
sorry, you might wanna undo this, and similar this things in the file

Konamiman replied:
I know ☹

vedanshujain reviewed (commented):
I added a couple of comments, but overall looks good. Neat work!

Konamiman committed:
OrderCacheController: backup enable option is now stored in memory.

Konamiman committed:
Convert conditions to Yoda :-(

Konamiman committed:
Add missing $

peterfabian commented:
Can we please test this with at least one of the in-memory caches like Redis to make sure it works before we merge it? Thanks!

Konamiman committed:
Merge branch 'trunk' into add/order_cache

Konamiman committed:
Merge branch 'trunk' into add/order_cache

Konamiman committed:
Use the new features engine to declare the cache as an experimental feature

Also decouple the orders cache mechanism from the COT feature,
it can be now used indepently of the COT feature and independently
of whether the new orders table is in use or not.

Konamiman committed:
Merge branch 'trunk' into add/order_cache

vedanshujain committed:
Merge branch 'trunk' into add/order_cache

vedanshujain committed:
Removed unused import and transient class.

vedanshujain reviewed (approved):
LGTM!
